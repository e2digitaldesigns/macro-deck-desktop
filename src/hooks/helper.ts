import _findIndex from "lodash/findIndex";
import { useGlobalData } from "./../hooks";

export interface IntUseHelperHook {
  getProfileIndex: () => number;
  getPageIndex: (_id: string) => number;
  getButtonPadIndex: (_id: string) => number;
  getActionIndex: (_id: string) => number;
}

const useHelperHook = (): IntUseHelperHook => {
  const globalData: any = useGlobalData();

  const getProfileIndex = (): number => {
    const index = _findIndex(globalData?.state?.profiles, (f: any) => {
      return f._id === globalData?.state?.activeProfile?._id;
    });

    return index;
  };

  const getPageIndex = (_id: string): number => {
    const profileIndex = getProfileIndex();
    const id =
      _id === "null" ? globalData?.state?.activeProfile?.page?._id : _id;

    const pageIndex = _findIndex(
      globalData?.state?.profiles?.[profileIndex].pages,
      (f: any) => {
        return f._id === id;
      }
    );

    return pageIndex;
  };

  const getButtonPadIndex = (_id: string): number => {
    const profileIndex = getProfileIndex();
    const pageIndex = getPageIndex("null");

    const id =
      _id === "null" ? globalData?.state?.activeProfile?.buttonPad?._id : _id;

    const buttonPadIndex = _findIndex(
      globalData?.state.profiles?.[profileIndex]?.pages[pageIndex].buttonPads,
      (f: any) => {
        return f._id === id;
      }
    );

    return buttonPadIndex;
  };

  const getActionIndex = (_id: string): number => {
    const profileIndex = getProfileIndex();
    const pageIndex = getPageIndex("null");
    const buttonPadIndex = getButtonPadIndex("null");

    const id =
      _id === "null" ? globalData?.state?.activeProfile?.action?._id : _id;

    const actionIndex = _findIndex(
      globalData?.state.profiles?.[profileIndex]?.pages[pageIndex]?.buttonPads[
        buttonPadIndex
      ]?.actions,
      (f: any) => {
        return f._id === id;
      }
    );

    return actionIndex;
  };

  return {
    getProfileIndex,
    getPageIndex,
    getButtonPadIndex,
    getActionIndex
  };
};

export default useHelperHook;
