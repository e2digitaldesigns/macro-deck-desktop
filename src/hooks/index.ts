export { default as useGlobalData } from "./useGlobalData";
export { default as useActions } from "./useActionsHook";
export { default as useButton } from "./useButtonHook";
export { default as useElectron } from "./useElectronHook";
export { default as useHelper } from "./helper";
export { default as usePage } from "./usePageHook";
export { default as useProfile } from "./useProfileHook";
