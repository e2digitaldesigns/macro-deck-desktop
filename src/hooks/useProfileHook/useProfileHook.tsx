import _cloneDeep from "lodash/cloneDeep";
import _filter from "lodash/filter";
import _find from "lodash/find";
import _findIndex from "lodash/findIndex";
import _sortBy from "lodash/sortBy";

import { useGlobalData } from "..";
import {
  IntProfile,
  IntButtonPads,
  IntActions,
  IntGlobalData,
  IntPages,
  IntGlobalContextInterface
} from "../../types";

import { useObj } from "../../hooks";

export interface IntEditState {
  editing: boolean;
  profileName: string;
  buttonPads: number;
}

export interface IntUseProfileHook {
  activateProfile: (_id: string) => void;
  createProfile: () => void;
  readProfiles: () => IntProfile[];
  readProfile: () => IntProfile;
  updateProfile: (_id: string, profileState: IntEditState) => void;
  deleteProfile: (_id: string) => void;
}

const useProfileHook = (): IntUseProfileHook => {
  const globalData: IntGlobalData = useGlobalData();
  const { pageObj, profileObj } = useObj();

  const activateProfile: IntUseProfileHook["activateProfile"] = (_id): void => {
    const state: IntGlobalContextInterface = _cloneDeep(globalData.state);
    const pages = _sortBy(
      _filter(state?.pages, (f: IntPages) => f.profileId === _id),
      "order"
    );

    state.active.profileId = _id;
    state.active.pageId = pages?.[0]?._id;
    state.active.buttonPadId = "";
    state.active.actionId = "";
    globalData.setState(state);
  };

  const createProfile = () => {
    const state = _cloneDeep(globalData.state);
    const profile = profileObj();
    const page = pageObj();
    page.profileId = profile._id;
    state.profiles.push(profile);
    state.pages.push(page);
    state.active.profileId = profile._id;
    state.active.pageId = page._id;
    globalData.setState(state);
  };

  const readProfiles = (): IntProfile[] | any => {
    const state = _cloneDeep(globalData.state);
    const profiles = state.profiles;
    return profiles;
  };

  const readProfile = (): IntProfile | any => {
    const state = _cloneDeep(globalData.state);

    const profile = _find(
      state.profiles,
      (f: IntProfile) => f._id === state.active.profileId
    );

    return profile;
  };

  const updateProfile = (_id: string, profileState: IntEditState): void => {
    const state = _cloneDeep(globalData.state);
    const profileIndex = _findIndex(
      state.profiles,
      (f: IntProfile) => f._id === _id
    );

    state.profiles[profileIndex].profileName = profileState.profileName;
    state.profiles[profileIndex].buttonPads = Number(profileState.buttonPads);
    globalData.setState(state);
  };

  const deleteProfile: IntUseProfileHook["deleteProfile"] = (
    _id: string
  ): void => {
    const state: IntGlobalContextInterface = _cloneDeep(globalData.state);
    state.profiles = _filter(state.profiles, (f: IntProfile) => f._id !== _id);
    if (state.profiles.length === 0) return;

    state.pages = _filter(state.pages, (f: IntPages) => f.profileId !== _id);

    state.buttonPads = _filter(
      state.buttonPads,
      (f: IntButtonPads) => f.profileId !== _id
    );

    state.actions = _filter(
      state.actions,
      (f: IntActions) => f.profileId !== _id
    );

    if (state.active.profileId === _id) {
      state.active.profileId = "";
      state.active.pageId = "";
      state.active.buttonPadId = "";
      state.active.actionId = "";
    }

    globalData.setState(state);
  };

  return {
    activateProfile,
    createProfile,
    readProfiles,
    readProfile,
    updateProfile,
    deleteProfile
  };
};

export default useProfileHook;
