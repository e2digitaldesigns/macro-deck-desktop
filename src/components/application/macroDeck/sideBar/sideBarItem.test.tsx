import React from "react";
import "@testing-library/jest-dom";
import { render, fireEvent, waitFor } from "@testing-library/react";
import SideBarItems from "./sideBarItem";
import * as hooks from "../../../../hooks";
import * as useProfileHook from "../../../../hooks/useProfileHook";

const mockData = {
  _id: "the-id",
  profileName: "the profile name",
  buttonPads: 24
};

const mockState = {
  userInformation: {
    _id: "",
    name: ""
  },
  templateInformation: {
    sideBarState: ""
  },
  settings: {},
  active: {
    profileId: "",
    buttonPadId: "",
    actionId: "",
    pageId: ""
  },
  profiles: [],
  pages: [],
  buttonPads: []
};

const mockUseProfile = {
  deleteProfile: () => jest.fn(),
  activateProfile: () => jest.fn(),
  updateProfile: () => jest.fn()
};

jest.mock("../../../hooks", () => ({
  useProfile: () => ({
    deleteProfile: () => jest.fn(),
    activateProfile: () => jest.fn(),
    updateProfile: () => jest.fn()
  }),
  useGlobalData: () => mockState
}));

const preFix = "template-sidebar-item__";

const testSetup = () => {
  return render(<SideBarItems profile={mockData} />);
};

let wrapper: any = null;
beforeEach(() => {
  wrapper = testSetup();
});

afterEach(() => {
  wrapper = null;
  jest.clearAllMocks();
});

describe("<Template Header Component/>", () => {
  it("Should render without errors", () => {
    const component = wrapper.getByTestId(preFix + "component");
    expect(component).toBeTruthy();
    expect(component).toMatchSnapshot();
  });

  it("Should toggle menu-item-active class", () => {
    const component = wrapper.getByTestId(preFix + "component");
    expect(component).toHaveClass("menu-item-inactive");

    const menuItemButton = wrapper.getByTestId(
      preFix + "component-menu-item-button"
    );
    expect(menuItemButton).toBeTruthy();

    fireEvent.click(menuItemButton);

    waitFor(() => {
      expect(component).toHaveClass("menu-item-active");
    });
  });

  it("Should delete profile", () => {
    const component = wrapper.getByTestId(preFix + "component");
    const deleteItemButton = wrapper.getByTestId(preFix + "button-remove");
    expect(deleteItemButton).toBeTruthy();

    fireEvent.click(deleteItemButton);
  });

  it("Should render header information", () => {
    const component = wrapper.getByTestId(preFix + "button-text");

    expect(component).toBeTruthy();
    expect(component).toHaveTextContent(
      `${mockData.profileName} (${mockData.buttonPads})`
    );
  });

  it("Should render ProfileName values", () => {
    const component = wrapper.getByTestId(preFix + "input_profile-name");

    expect(component).toBeTruthy();
    expect(component.value).toBe(mockData.profileName);
  });

  it("Should render ButtonPad values", () => {
    const component = wrapper.getByTestId(preFix + "select-button-pads");

    expect(component).toBeTruthy();
    expect(component.value).toBe(mockData.buttonPads.toString());
  });

  it("Should render submit button", () => {
    const submitButton = wrapper.getByTestId(preFix + "button-submit");
    expect(submitButton).toBeTruthy();
  });

  it("Update should succeed", () => {
    const spy = jest.spyOn(hooks, "useProfile");

    const submitButton = wrapper.getByTestId(preFix + "button-submit");
    const newNameGood = "New Profile Name XXXXX XXXXX XXXXX";
    // const newPadCount = 12;

    const input = wrapper.getByTestId(preFix + "input_profile-name");
    fireEvent.change(input, { target: { value: newNameGood } });
    expect(input.value).toBe(newNameGood);

    // const select = wrapper.getByTestId(preFix + "select-button-pads");
    // fireEvent.change(select, { target: { value: newPadCount } });
    // expect(select.value).toBe(newPadCount.toString());

    fireEvent.click(submitButton);
    // expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });
});

it("Update should fail", () => {
  const submitButton = wrapper.getByTestId(preFix + "button-submit");
  const newNameBad = "Ne";

  const input = wrapper.getByTestId(preFix + "input_profile-name");
  fireEvent.change(input, { target: { value: newNameBad } });
  expect(input.value).toBe(newNameBad);

  fireEvent.click(submitButton);

  waitFor(() => {
    expect(input).toHaveFocus();
  });
});
