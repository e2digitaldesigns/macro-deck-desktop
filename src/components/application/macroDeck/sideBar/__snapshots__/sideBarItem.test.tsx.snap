// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideBarItem/> Should render without errors 1`] = `
<div
  class="menu-item menu-item-inactive"
  data-testid="template-sidebar-item__component"
>
  <div
    class="menu-item-button"
    data-testid="template-sidebar-item__component-menu-item-button"
  >
    <div
      class="menu-item-button-text"
      data-testid="template-sidebar-item__button-text"
    >
      001 the profile name
       (
      12
      )
    </div>
    <div
      class="menu-item-button-edit"
      data-testid="template-sidebar-item__button-edit"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
        data-testid="EditIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
    </div>
    <div
      class="menu-item-button-remove"
      data-testid="template-sidebar-item__button-remove"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
        data-testid="DeleteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </div>
  </div>
  <div
    class="menu-item-information false"
    data-testid="template-sidebar-item__menu-item-information-wrapper"
  >
    <div
      class="menu-item-infomation-inner"
    >
      Name:
       
      <input
        data-testid="template-sidebar-item__input_profile-name"
        name="profileName"
        type="text"
        value="001 the profile name"
      />
      <br />
      ButtonPads:
      <select
        data-testid="template-sidebar-item__select-button-pads"
        name="buttonPads"
      >
        <option
          value="6"
        >
          6
        </option>
        <option
          value="8"
        >
          8
        </option>
        <option
          value="12"
        >
          12
        </option>
        <option
          value="15"
        >
          15
        </option>
        <option
          value="24"
        >
          24
        </option>
        <option
          value="32"
        >
          32
        </option>
      </select>
      <br />
      <button
        class="menu-item-information-submit"
        data-testid="template-sidebar-item__button-submit"
      >
        submit
      </button>
    </div>
  </div>
</div>
`;
